### PR TITLE
Add TF2ModelImpl.

### PR DESCRIPTION
**Pull request by jsheng-jian**
TF2ModelImpl is a wrapper on top of tensorflow2.x which implements DLRModel API.
It allows to load tensorflow2.x saved model and run inference. I have verified with `ssd_mobilenet_v2_320x320_coco17_tpu-8` too.

```
root@ce5f72a26aac:/neo-ai-dlr# python test_tf2.py

 CALL HOME FEATURE ENABLED
                            

 You acknowledge and agree that DLR collects the following metrics to help improve its performance.                             
 By default, Amazon will collect and store the following information from your device:                             

 record_type: <enum, internal record status, such as model_loaded, model_>,                             
 arch: <string, platform architecture, eg 64bit>,                             
 osname: <string, platform os name, eg. Linux>,                             
 uuid: <string, one-way non-identifable hashed mac address, eg. 8fb35b79f7c7aa2f86afbcb231b1ba6e>,                             
 dist: <string, distribution of os, eg. Ubuntu 16.04 xenial>,                             
 machine: <string, retuns the machine type, eg. x86_64 or i386>,                             
 model: <string, one-way non-identifable hashed model name, eg. 36f613e00f707dbe53a64b1d9625ae7d>                             

 If you wish to opt-out of this data collection feature, please follow the steps below:                             
        1. Disable it with through code:                             
                 from dlr.counter.phone_home import PhoneHome                             
                 PhoneHome.disable_feature()                            
        2. Or, create a config file, ccm_config.json inside your DLR target directory path, i.e. python3.6/site-packages/dlr/counter/ccm_config.json. Then added below format content in it, {"enable_phone_home" : false}                             
        3. Restart DLR application.                             
        4. Validate this feature is disabled by verifying this notification is no longer displayed, or programmatically with following command:                             
                from dlr.counter.phone_home import PhoneHome                             
                PhoneHome.is_enabled() # false as disabled 
2022-03-09 05:45:35.948019: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:35.955873: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:35.956537: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:35.957721: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-03-09 05:45:35.958554: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:35.959198: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:35.959813: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:36.586782: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:36.587450: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:36.588067: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-03-09 05:45:36.588649: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1510] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 13813 MB memory:  -> device: 0, name: Tesla T4, pci bus id: 0000:00:1e.0, compute capability: 7.5
WARNING:tensorflow:Skipping full serialization of Keras layer <__main__.TestModel object at 0x7ff8da343ac8>, because it is not built.
2022-03-09 05:45:37,061 INFO found TF2.x saved model, dlr will use TensorFlow runtime.
2022-03-09 05:45:37.191709: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
All tests passed!
```

### CODE REVIEW

File: python/dlr/api.py [hunk 1]
```diff
@@ -43,6 +43,15 @@ def _find_model_file(model_path, ext):
         return model_files[0]
     return None
 
+def _find_saved_model(model_path):
```
Comment (apivovarov) [inline python/dlr/api.py:46]: 1. Lets add tf2 to the method name
2. Looks like `_find_model_file` is not used anymore

File: python/dlr/tf_model.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+from .neologger import create_logger
+from .metadata import VERSION
+
+class TFModelImpl(IDLRModel):
+    """
+    TFModelImpl is a wrapper on top of tensorflow which implements DLRModel API
+    Parameters
+    ----------
+    model_path : str
+        Full path to the directory containing the saved model
```
Comment (apivovarov) [inline python/dlr/tf_model.py:14]: It sounds like a parent dir of the saved model. Lets says "Full path to the saved model directory"

File: python/dlr/api.py [hunk 3]
```diff
@@ -43,6 +43,15 @@ def _find_model_file(model_path, ext):
...
+    if not os.path.isdir(model_path):
+        return False
+    model_file = glob.glob(os.path.abspath(os.path.join(model_path, '*' + '.pb')))
```
Comment (apivovarov) [inline python/dlr/api.py:49]: TF2 tf.saved_model.load looks for `saved_model.pbtxt` or `saved_model.pb`.

File: tests/python/unittest/test_tf2.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+def test_tf_model_on_cpu_0():
+    test_tf_model("cpu", 0)
+
+def test_tf_model_on_gpu_0():
+    test_tf_model("gpu", 0)
+
+if __name__ == '__main__':
+    test_tf_model()
+    print('All tests passed!')
```
Comment (apivovarov) [inline tests/python/unittest/test_tf2.py:61]: No new line at the end of the file. Check your editor/IDE settings to include new line at the end of the files.

File: python/dlr/tf_model.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
+import tensorflow as tf
```
Comment (apivovarov) [inline python/dlr/tf_model.py:1]: 1. Test file name is `test_tf2.py`. Lets rename `tf_model.py` to `tf2_model.py` too.
2. Should we check that tf version is 2.x?

File: python/dlr/tf_model.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    def __init__(self, model_path, dev_type=None, dev_id=None, error_log_file=None, use_default_dlr=False):
+
+        self.model_path = model_path
+        self.logger = create_logger(log_file=error_log_file)
+        if dev_type is not None:
+            devices = ["cpu", "gpu"]
+            if dev_type not in devices:
+                raise ValueError(
+                    "Invalid device type {}. Valid devices: {}".format(dev_type, devices))
+        if dev_type == 'gpu':
```
Comment (apivovarov) [inline python/dlr/tf_model.py:30]: We do not ask TF to run the saved_model on a particular Device.  Looks like `dev_type` and `dev_id` params are meaningless for TF2 adapter. 
It should be ok for Jetson users to skip `dev_type` and `dev_id` params. Example:
```
model=dlr.DLRModel("./saved_tf_trt_model")
```
Lets `list_physical_devices('GPU')` and `set_memory_growth(True)` for all of them regardless of `dev_type` and `dev_id`.

2. If user tries to set `dev_type` and `dev_id` we should show warn message saying that `dev_type` and `dev_id` are not supported for TF2 Models and the params are ignores.

File: python/dlr/tf_model.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            self.logger.info("TF memory_growth: {}".format(
+                tf.config.experimental.get_memory_growth(physical_devices[dev_id])))
+
+        self.use_default_dlr = use_default_dlr
+        self._lib = None
+        self.version = VERSION
+
+        self.saved_model = tf.saved_model.load(model_path)
+        tag_set = 'serve'
+        assert len(self.saved_model.signatures) > 0
```
Comment (apivovarov) [inline python/dlr/tf_model.py:43]: Lets add assertion_message

File: python/dlr/tf_model.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        if isinstance(input_values, dict):
+            for k in input_values.keys():
+                if not isinstance(k, str):
+                    raise ValueError("input key must be string")
+                self._validate_input_name(k)
+            self.input_values = input_values
+        elif isinstance(input_values, (list, tuple)):
+            l_names, l_input_values = len(
+                self.input_tensor_names), len(input_values)
+            assert l_names == l_input_values, "Input tensor number does not match, expect {}, got {}".format(
```
Comment (apivovarov) [inline python/dlr/tf_model.py:106]: `Wrong number of inputs, expect {}, got {}`

File: python/dlr/tf_model.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            return None
+        out = self.input_values[name]
+        if shape is not None:
+            out = out.reshape(shape)
+        return out
+
+    def _validate_input_name(self, name):
+        if name not in self.input_tensor_names:
+            raise ValueError(
+                "Invalid input tensor name '{}'. List of input tensor names: {}".format(name, self.input_tensor_names))
```
Comment (apivovarov) [inline python/dlr/tf_model.py:94]: `Invalid input name '{}'. List of input names: {}`

File: python/dlr/api.py (hunk 1 above)
Comment (jsheng-jian) [reply python/dlr/api.py]: 1. done
2. removed `_find_model_file`

File: python/dlr/tf_model.py (hunk 2 above)
Comment (jsheng-jian) [reply python/dlr/tf_model.py]: done

File: python/dlr/api.py (hunk 3 above)
Comment (jsheng-jian) [reply python/dlr/api.py:38]: done

File: tests/python/unittest/test_tf2.py (hunk 4 above)
Comment (jsheng-jian) [reply tests/python/unittest/test_tf2.py]: done

File: python/dlr/tf_model.py (hunk 5 above)
Comment (jsheng-jian) [reply python/dlr/tf_model.py]: 1. done
2. added `assert version.parse(tf.__version__) >= version.parse(TENSORFLOW_VERSION)` to check tensorflow version

File: python/dlr/tf_model.py (hunk 6 above)
Comment (jsheng-jian) [reply python/dlr/tf_model.py]: done

File: python/dlr/tf_model.py (hunk 7 above)
Comment (jsheng-jian) [reply python/dlr/tf_model.py]: done

File: python/dlr/tf_model.py (hunk 8 above)
Comment (jsheng-jian) [reply python/dlr/tf_model.py]: done

File: python/dlr/tf_model.py (hunk 9 above)
Comment (jsheng-jian) [reply python/dlr/tf_model.py]: done

**Review by apivovarov — APPROVED** [review]

**Pull request merged**